plugins/auth-backend/src/providers/bitbucketServer/provider.test.ts

```diff
@@ -0,0 +1,377 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as helpers from '../../lib/passport/PassportStrategyHelper';
+import { makeProfileInfo } from '../../lib/passport/PassportStrategyHelper';
+import { AuthResolverContext } from '../types';
+import {
+  BitbucketServerAuthProvider,
+  BitbucketServerOAuthResult,
+} from './provider';
+import { commonByEmailResolver } from '../resolvers';
+
+jest.mock('../../lib/passport/PassportStrategyHelper', () => {
+  return {
+    ...jest.requireActual('../../lib/passport/PassportStrategyHelper'),
+    executeFrameHandlerStrategy: jest.fn(),
+    executeRefreshTokenStrategy: jest.fn(),
+    executeFetchUserProfileStrategy: jest.fn(),
+  };
+});
+
+const mockFrameHandler = jest.spyOn(
+  helpers,
+  'executeFrameHandlerStrategy',
+) as unknown as jest.MockedFunction<
+  () => Promise<{
+    result: BitbucketServerOAuthResult;
+    privateInfo: { refreshToken?: string };
+  }>
+>;
+
+const passportProfile = {
+  id: '123',
+  username: 'john.doe',
+  provider: 'bitubcketServer',
+  displayName: 'John Doe',
+  emails: [{ value: 'john@doe.com' }],
+  photos: [{ value: 'https://bitbucket.org/user/123/avatar' }],
+};
+
+const mockFetchUserRequests = (
+  failOnWhoAmI: boolean = false,
+  whoAmIValue: string = passportProfile.username,
+  failOnGetUser: boolean = false,
+  getUserOk: boolean = true,
+  avatarUrl: string = '/user/123/avatar',
+  setDisplayName: boolean = true,
+  setUserName: boolean = true,
+) => {
+  const fetchMock = global.fetch as jest.Mock;
+  if (failOnWhoAmI) {
+    fetchMock.mockRejectedValueOnce(() => {});
+  } else {
+    fetchMock.mockResolvedValueOnce({
+      headers: { get: jest.fn(() => whoAmIValue) },
+    });
+  }
+  if (failOnGetUser) {
+    fetchMock.mockRejectedValueOnce(() => {});
+  } else {
+    fetchMock.mockResolvedValueOnce({
+      ok: getUserOk,
+      json: () => ({
+        name: setUserName ? 'john.doe' : undefined,
+        emailAddress: 'john@doe.com',
+        id: 123,
+        displayName: setDisplayName ? 'John Doe' : undefined,
+        active: true,
+        slug: 'john.doe',
+        type: 'NORMAL',
+        links: {
+          self: [
+            {
+              href: 'https://bitbucket.org/users/john.doe',
+            },
+          ],
+        },
+        avatarUrl: avatarUrl,
+      }),
+    });
+  }
+};
+
+describe('BitbucketServerAuthProvider', () => {
+  const originalFetch = global.fetch;
```

Comment:
We pervasively make use of `msw` to stub out network traffic, making things fast and robust. It also protects against accidentally leaking real traffic outside of the test.

https://github.com/backstage/backstage/blob/3cf2190c2df803833b7f6f057d873bb6e96bcb46/plugins/scaffolder/src/api.test.ts#L28

Reply:
Done. I never worked with `msw` before, so I hope everything is fine. 

Btw, thanks for the suggestion - `msw` looks like a great tool :)